frame/stake-tracker/src/lib.rs

```diff
@@ -0,0 +1,228 @@
+		}
+
+		if T::Staking::is_validator(&current_stake.stash) {
+			update_approval_stake(&current_stake.stash);
+
+			let _ = T::VoterList::on_update(&current_stake.stash, Self::to_vote(current_active))
+				.defensive_proof("any validator should have an entry in the voter list.");
+		}
+	}
+
+	fn on_nominator_add(who: &T::AccountId, prev_nominations: Vec<T::AccountId>) {
+		let nominations = T::Staking::nominations(who).unwrap_or_default();
+		let new = nominations.iter().filter(|n| !prev_nominations.contains(&n));
+		let obsolete = prev_nominations.iter().filter(|n| !nominations.contains(&n));
+
+		let update_approval_stake = |nomination: &T::AccountId, new_stake: BalanceOf<T>| {
+			ApprovalStake::<T>::set(&nomination, Some(new_stake));
+
+			if T::TargetList::contains(&nomination) {
+				let _ = T::TargetList::on_update(&nomination, new_stake).defensive();
+			}
+		};
+
+		for nomination in new {
+			// Create a new entry if it does not exist
+			let new_stake = Self::approval_stake(&nomination)
+				.unwrap_or_default()
+				.saturating_add(Self::slashable_balance_of(who));
+
+			update_approval_stake(&nomination, new_stake);
+		}
+
+		for nomination in obsolete {
+			// we should actually fail if an old nomination was not in the map, which should never
+			// happen.
+			let new_stake = Self::approval_stake(&nomination)
+				.unwrap()
+				.saturating_sub(Self::slashable_balance_of(who));
+
+			update_approval_stake(&nomination, new_stake);
+		}
+		let _ =
+			T::VoterList::on_insert(who.clone(), Self::to_vote(Self::slashable_balance_of(who)))
+				.defensive();
+	}
+
+	/// This should only be called if that stash isn't already a validator. Note, that if we want to
+	/// properly track ApprovalStake here - we need to make sure we subtract the validator stash
+	/// balance when they chill?
+	///	Why? Because we don't remove ApprovalStake when a validator chills and we need to make sure
+	/// their self-stake is up-to-date and not applied twice.
+	fn on_validator_add(who: &T::AccountId) {
+		let self_stake = Self::slashable_balance_of(who);
+		let new_stake = Self::approval_stake(who).unwrap_or_default().saturating_add(self_stake);
+
+		// maybe update sorted list.
+		let _ = T::VoterList::on_insert(who.clone(), Self::to_vote(self_stake)).defensive();
+
+		// TODO: Make sure this always works. Among other things we need to make sure that when the
+		// migration is run we only have active validators in TargetList and not the chilled ones.
+		let _ = T::TargetList::on_insert(who.clone(), new_stake).defensive();
+		ApprovalStake::<T>::set(who, Some(new_stake));
+	}
+
+	fn on_validator_remove(who: &T::AccountId) {
+		let _ = T::TargetList::on_remove(who).defensive();
+		let _ = T::VoterList::on_remove(who).defensive();
+
+		// This will panic if the validator is not in the map, but this should never happen!
+		ApprovalStake::<T>::mutate(who, |x: &mut Option<BalanceOf<T>>| {
+			x.map(|b| b.saturating_sub(Self::slashable_balance_of(who)))
```

Comment:
`mutate` expects you to mutate `x`. what you are doing here is returning some value from the closure, which will not update `x` at all. If you had at least one test, this would have been caught by now ;)

Reply:
ah yeah, it should assign to `*x`.